### PR TITLE
Feature/zcs 3825

### PR DESCRIPTION
**Pull request by grishick**
For ZCS-3825 add id range query to zimbra query syntax and add unit tests for ItemQuery.
Also added IMAP tests from another branch and kept the failing tests ignored for remote IMAP case.

### CODE REVIEW

File: store/src/java/com/zimbra/qa/unittest/SharedImapTests.java [hunk 1]
```diff
@@ -2565,4 +2565,102 @@ public void appendTooLarge() throws Exception {
...
+
+        /*
+         * The reason for performing all assertions together and reporting all numbers in each assertion message is to get a more complete picture
+         * when tests fail.
+         */
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 1:* UNDELETED'. Results are: '1:*' : %d, '10:800' : %d, '1:2000' : %d, '1000:1999' : %d", 
+                oneStarResults, ten8HundredResults, oneTwoThousandResults, oneThousand3NineResults), numMessages, oneStarResults);
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 10:800 UNDELETED'. Results are: '1:*' : %d, '10:800' : %d, '1:2000' : %d, '1000:1999' : %d", 
+                oneStarResults, ten8HundredResults, oneTwoThousandResults, oneThousand3NineResults), 791, ten8HundredResults);
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 10:2000 UNDELETED'. Results are: '1:*' : %d, '10:800' : %d, '1:2000' : %d, '1000:1999' : %d",
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/SharedImapTests.java]: s/10:2000/1:2000/

File: store/src/java/com/zimbra/qa/unittest/SharedImapTests.java [hunk 2]
```diff
@@ -2565,4 +2565,102 @@ public void appendTooLarge() throws Exception {
...
+         * when tests fail.
+         */
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 1:* UNDELETED'. Results are: '1:*' : %d, '10:800' : %d, '1:2000' : %d, '1000:1999' : %d", 
+                oneStarResults, ten8HundredResults, oneTwoThousandResults, oneThousand3NineResults), numMessages, oneStarResults);
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 10:800 UNDELETED'. Results are: '1:*' : %d, '10:800' : %d, '1:2000' : %d, '1000:1999' : %d", 
+                oneStarResults, ten8HundredResults, oneTwoThousandResults, oneThousand3NineResults), 791, ten8HundredResults);
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 10:2000 UNDELETED'. Results are: '1:*' : %d, '10:800' : %d, '1:2000' : %d, '1000:1999' : %d", 
+                oneStarResults, ten8HundredResults, oneTwoThousandResults, oneThousand3NineResults), 2000, oneTwoThousandResults);
+        assertEquals(String.format("Wrong number of results for 'UID SEARCH 10:2000 UNDELETED'. Results are: '1:*' : %d, '1000:1999' : %d, '1:2000' : %d, '1000:1999' : %d",
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/SharedImapTests.java]: s/10:2000/1000:1999/

File: store/src/java/com/zimbra/qa/unittest/TestImapViaEmbeddedRemote.java [hunk 3]
```diff
@@ -36,4 +40,32 @@ public void tearDown() throws ServiceException, DocumentException, ConfigExcepti
     protected int getImapPort() {
         return imapServer.getImapBindPort();
     }
+
+    @Override
+    @SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")  // checking done in called methods
+    @Ignore("ZCS-1856 - fetch immediately after append doesn't find the item")
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestImapViaEmbeddedRemote.java]: ZCS-1856 should be fixed now.  Is your repo's view of develop out of date?

File: store/src/java/com/zimbra/qa/unittest/TestImapViaImapDaemon.java [hunk 4]
```diff
@@ -308,4 +313,32 @@ public void testFlushNonImapCacheTypes() throws Exception {
               }
           }
       }
+
+    @Override
+    @SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")  // checking done in called methods
+    @Ignore("ZCS-1856 - fetch immediately after append doesn't find the item")
```
Comment (fatkudu) [inline store/src/java/com/zimbra/qa/unittest/TestImapViaImapDaemon.java]: Again, shouldn't be needed now.

**Comment by grishick** [comment]
@fatkudu this PR does not implement search by UIDs, only by item IDs in the database.

File: store/src/java/com/zimbra/cs/index/query/ItemQuery.java [hunk 5]
```diff
@@ -37,17 +37,32 @@
...
+        boolean rangeQuery = false;
         List<ItemId> itemIds = new ArrayList<ItemId>();
 
         if (str.equalsIgnoreCase("all")) {
             allQuery = true;
         } else if (str.equalsIgnoreCase("none")) {
             noneQuery = true;
+        } else if(str.indexOf("--") > 0) {
+            String[] items = str.split("--");
+            for (int i = 0; i < items.length && i < 2; i++) {
```
Comment (ilyusha) [inline store/src/java/com/zimbra/cs/index/query/ItemQuery.java]: Does this mean that a query like "100--200--300" would be treated like a range from 100 to 200? Is it worth it to check for invalid range inputs like this?

Comment (grishick) [reply store/src/java/com/zimbra/cs/index/query/ItemQuery.java]: good catch, @ilyusha ! Yes, I should add a syntax check here.

**Pull request merged**